vtex/faststore
# chore: Simplify search

## What's the purpose of this pull request?
This PR simplifies the Search module. The simplification occurs by moving lots of generic code from `base.store` to the SDK while making the search module itself more simple and tight integrated. 

## How it works? 
Before, the search module in the SDK was split into:
1. State: an object containing the state of the search, like selected facets, the search term, sorting etc
2. Serializer: to parse/format the search state into a url
3. Reducer: functions to change the state, like setting a new search term, selecting sort and toggling facets. 

All these functions were pure and didn't really depended on React. This was good because we could make it work outside React, however, made lots of React glue code be deferred to `base.store`, polluting the code base and generating unnecessary complexity. 
This PR moves this glue code from `base.store` into the SDK and makes the State,Serializer,Reducer depend explicitly from React so the final code is simpler and better integrated with React. 

The State/Serializer code was implemented using React's `useReducer` hook into the `useSearchState.ts` file. This file manages the search state, like which term is selected, which sort is applied and what are the facets selected. 
To help store implementors develop better UI's with infinite scroll, `useInfiniteSearchState.ts` hook was created. This hook manages the amount of pages in the infinite scroll and helps appending/prepending new pages to the list.
To store both search state and infinite scroll state, a `SearchContext` was created. It uses both `useSearchState.ts` and `useInfiniteSearchState.ts` hooks and is the only context exported. 
To help store devs nail on SEO, a `usePagination.ts` is exported so devs can easily get links for next/prev routes prefetch routes.
Finally, to help parsing/format urls, the serializer from previous versions was kept and is still exported

## How to test it?
Make sure to open base.store's deploy preview and make sure all seach-related features are still there, i.e. the facets,sorting are still working

### `base.store` Deploy Preview
https://github.com/vtex-sites/base.store/pull/115
